### PR TITLE
[Auto] Fix #195845: [inductor] mix-order reduction: split column sum ignores xmask, wrong results wh

### LINKED ISSUE #195845: [inductor] mix-order reduction: split column sum ignores xmask, wrong results when rows % XBLOCK != 0
With `triton.mix_order_reduction` on (OSS default), a row and a column reduction over the same input fuse into one kernel. The column reduction is accumulated per RSPLIT chunk with `tl.sum(<value, 0)` without masking the tile by `xmask`. Padded rows load `other=0.0`, so any fused prologue that maps 0 to nonzero (`+ bias`, `exp`, softmax, ...) is summed once per padded row whenever `rows % XBLOCK != 0`.

```python
import torch

def f(x):
    y = x * 2 + 0.25
    return y.max(dim=-1).values, y.float().sum(dim=0)

x = torch.randn(40961, 129, device="cuda", dtype=torch.bfloat16)
_, colsum = torch.compile(f)(x)
ref = (x.float() * 2 + 0.25).sum(dim=0)
print((colsum - ref).abs().max().item())  # 0.75 (3 padded rows x 0.25); 0.0 for 40960 rows
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by mlazos**
Human direction from @mlazos:

> "go ahead and do all of the updates, I'll review on github independently"

> **AI-assisted draft for that independent GitHub review:**
>
> ## Summary
>
> Fixes #195845.
>
> Correct Triton mix-order split column reductions when the row count is not a multiple of `XBLOCK`. The change also adds direct code-generation coverage for sum and product identity masking and a device-generic numerical regression for CUDA and XPU.
>
> ## Root cause
>
> Mix-order reduction performs each split's partial reduction across the X dimension after running the fused pointwise body. Although masked loads use a fill value, pointwise operations can transform that value into something other than the reduction identity. The partial reduction previously consumed those transformed values without applying `xmask` again, so padded rows could contribute to the result.
>
> ## Fix
>
> For nonconstant X masks, apply `xmask` immediately before each partial X-axis reduction and replace inactive lanes with the already-computed identity returned by `ir.Reduction.default_accumulator`. This keeps masked lanes neutral for each supported reduction type even when fused pointwise operations transform the original load fill value. The same `default` value initializes the accumulator and fills inactive lanes. Mix-order X dimensions also retain a dynamic mask when their logical size is one, because the selected `XBLOCK` can still be larger than one; when mask optimization is enabled, an explicit fixed `XBLOCK=1` remains mask-free.
>
> ## Architecture
>
> Applicability: Applicable
>
> Description: The change is localized to mix-order mask selection and the mix-order branch of `TritonKernel.codegen_body`. It enforces the invariant that every value entering a partial X-axis reduction is either produced by a valid X lane or replaced by that reduction's identity. A mask is omitted only when the chosen X block is proven to contain no padded lanes.
>
> Examples: A fused pointwise expression such as `x * 2 + 0.25` can turn a masked-load zero into `0.25`. The generated kernel now changes inactive lanes back to `0` before a sum, or to `1` before a product, before reducing across X.
>
> ## Example
>
> ```python
> import torch
>
>
> def f(x):
>     y = x * 2 + 0.25
>     return y.max(dim=-1).values, y.float().sum(dim=0)
>
> x = torch.zeros((40961, 129), dtype=torch.bfloat16, device="cuda")
> expected = f(x)
> actual = torch.compile(f)(x)
> torch.testing.assert_close(actual, expected)
> ```
>
> This illustrates the user-visible behavior; it does not by itself guarantee that the mix-order kernel is selected. The focused code-generation test directly constructs the exact mix-order path and checks its emitted mask. In the affected compiled pattern, padded X lanes could contribute transformed values to `actual[1]`; the fix replaces them with the sum identity immediately before the partial reduction.
>
> ## Testing
>
> These are reported local and CI outcomes. The local commands were run in the exact updated worktree unless explicitly qualified, but their raw terminal logs were not retained as independently auditable artifacts.
>
> - With CUDA hidden, the focused code-generation test passed all four combinations of `{sum, prod}` and `{xnumel=40961 with optimize_mask=False, xnumel=1 with optimize_mask=True}`. Reverting the full production mask change made all four cases fail because the identity-selecting `tl.where` was absent. Removing only the singleton-mask condition made the two singleton cases fail while both non-singleton cases continued to pass.
> - The focused test supplies fixed block sizes while constructing the kernel and its indexed expressions, avoiding runtime device-property queries on CPU-only builders. It then clears that construction-only override before body emission so the singleton case exercises the production non-fixed mask decision. Before the focused test supplied fixed block sizes, five CPU-only CI shards failed while trying to query CUDA device properties.
> - On CUDA, the full `TestCodegenTriton` class ran 53 tests: 48 passed and 5 skipped.
> - At the current head, the CUDA numerical regression passed on an NVIDIA H100. In an earlier run with the production masking fix temporarily reverted, all 129 elements of the column-sum output mismatched with a maximum absolute error of 0.75; restoring the fix made the same test pass.
> - XPU coverage is instantiated by the device-generic numerical test but was not run locally.
> - All three changed files passed two `lintrunner -a` autofix passes followed by a final plain `lintrunner` verification.
>
> Portable equivalents of the local invocations:
>
> ```bash
> CUDA_VISIBLE_DEVICES='' PYTHONPATH="$PWD" PYTHONDONTWRITEBYTECODE=1 python test/inductor/test_codegen_triton.py -k test_mix_order_partial_accumulate_masks_x -v
> CUDA_VISIBLE_DEVICES=0 PYTHONPATH="$PWD" PYTHONDONTWRITEBYTECODE=1 python test/inductor/test_codegen_triton.py TestCodegenTriton
> CUDA_VISIBLE_DEVICES=0 PYTORCH_TEST_WITH_SLOW=1 PYTHONPATH="$PWD" python test/inductor/test_mix_order_reduction.py MixOrderReductionNumericTestCUDA.test_split_column_reduction_masks_padded_rows_cuda -v
> # Autofix pass 1
> lintrunner -a test/inductor/test_codegen_triton.py test/inductor/test_mix_order_reduction.py torch/_inductor/codegen/triton.py
> # Autofix pass 2
> lintrunner -a test/inductor/test_codegen_triton.py test/inductor/test_mix_order_reduction.py torch/_inductor/codegen/triton.py
> # Final read-only verification
> lintrunner test/inductor/test_codegen_triton.py test/inductor/test_mix_order_reduction.py torch/_inductor/codegen/triton.py
> ```
>
> cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (architecture, correctness, simplification, generalization, and GPT-5.6 final approval). This is AI-generated — please verify findings manually.*

---

No verified issues found.

**Comment by mlazos** [comment]
## Automated feedback and CI resolution summary

> @mlazos, please review these automated resolutions. 
> Original review threads remain open until you decide to resolve them.

Latest processed commit: `cad5bd86e87e6e9a4a3bcda539341ca7bc436371`

| Source | Status | Resolution | Verification |
| --- | --- | --- | --- |
| [conversation:5615485642](https://github.com/pytorch/pytorch/pull/196565#issuecomment-5615485642) | already_addressed | The claimed review entry did not identify an actionable defect. Independent review preserved the production fix, expanded its numeric regression to the shared XPU backend, and documented the post-pointwise masking invariant. | Verified the production fix with red/green codegen and H100 numeric regressions; the fixed code passed 2 codegen cases and the 129-output numeric case, while removing it failed both checks. The full TestCodegenTriton class passed 51 tests with 5 skips, three five-pass review cycles ended with approval after resolving two test-quality findings, and all required lint passes reported no issues. |
| [check:106578964861](https://github.com/pytorch/pytorch/actions/runs/35674247153/job/106578964861) | fixed | Made the direct mix-order code-generation regression independent of runtime CUDA device discovery by supplying deterministic fixed block sizes. | At current head cad5bd86e87e, the exact focused test passes all four cases with CUDA hidden and the CUDA numerical regression passes on an NVIDIA H100. Reverting the production mask fix makes all four focused cases fail for the intended missing-identity-mask reason; an earlier implementation-reverted numerical run failed all 129 outputs with maximum absolute error 0.75. |
| [check:106579306507](https://github.com/pytorch/pytorch/actions/runs/35674247153/job/106579306507) | fixed | Made the direct mix-order code-generation regression independent of runtime CUDA device discovery by supplying deterministic fixed block sizes. | At current head cad5bd86e87e, the exact focused test passes all four cases with CUDA hidden and the CUDA numerical regression passes on an NVIDIA H100. Reverting the production mask fix makes all four focused cases fail for the intended missing-identity-mask reason; an earlier implementation-reverted numerical run failed all 129 outputs with maximum absolute error 0.75. |
| [check:106579307083](https://github.com/pytorch/pytorch/actions/runs/35674247153/job/106579307083) | fixed | Made the direct mix-order code-generation regression independent of runtime CUDA device discovery by supplying deterministic fixed block sizes. | At current head cad5bd86e87e, the exact focused test passes all four cases with CUDA hidden and the CUDA numerical regression passes on an NVIDIA H100. Reverting the production mask fix makes all four focused cases fail for the intended missing-identity-mask reason; an earlier implementation-reverted numerical run failed all 129 outputs with maximum absolute error 0.75. |
| [check:106579306789](https://github.com/pytorch/pytorch/actions/runs/35674247153/job/106579306789) | fixed | Made the direct mix-order code-generation regression independent of runtime CUDA device discovery by supplying deterministic fixed block sizes. | At current head cad5bd86e87e, the exact focused test passes all four cases with CUDA hidden and the CUDA numerical regression passes on an NVIDIA H100. Reverting the production mask fix makes all four focused cases fail for the intended missing-identity-mask reason; an earlier implementation-reverted numerical run failed all 129 outputs with maximum absolute error 0.75. |
| [check:106579306960](https://github.com/pytorch/pytorch/actions/runs/35674247153/job/106579306960) | fixed | Made the direct mix-order code-generation regression independent of runtime CUDA device discovery by supplying deterministic fixed block sizes. | At current head cad5bd86e87e, the exact focused test passes all four cases with CUDA hidden and the CUDA numerical regression passes on an NVIDIA H100. Reverting the production mask fix makes all four focused cases fail for the intended missing-identity-mask reason; an earlier implementation-reverted numerical run failed all 129 outputs with maximum absolute error 0.75. |

This comment is updated in place as new feedback is processed.
